### ISSUE TITLE
Add history_instance to pre_create_historical_record signal

### ISSUE THREAD

**Issue by dgilge** [treyhunner/django-simple-history#437]
**Problem Statement**
I'd like to set an extra field before saving the historical record (with the user's IP address and probably the form's URL). I think it's a good time to discuss that now because the signals aren't released yet and it was a not backwards compatible change.

**Describe the solution you'd like**
Change https://github.com/treyhunner/django-simple-history/blob/master/simple_history/models.py#L345-L364 to
```python
    def create_historical_record(self, instance, history_type):
        manager = getattr(instance, self.manager_name)
        history_instance = manager.model()
        history_instance.history_date = getattr(instance, '_history_date', now())
        history_instance.history_user = self.get_history_user(instance)
        history_instance.history_change_reason = getattr(instance, 'changeReason', None)

        pre_create_historical_record.send(
            sender=manager.model, instance=instance,
            history_instance=history_instance,  # this is the important difference
            history_date=history_instance.history_date,
            history_user=history_instance.history_user,
            history_change_reason=history_instance.history_change_reason,
        )

        for field in self.fields_included(instance):
            setattr(history_instance, field.attname, getattr(instance, field.attname))
        history_instance.save()
```

**Describe alternatives you've considered**
1. Override `create_historical_record`:

```python
    def create_historical_record(self, instance, history_type):
        history_date = getattr(instance, '_history_date', now())
        history_user = self.get_history_user(instance)
        history_change_reason = getattr(instance, 'changeReason', None)
        ip_address = instance.request.META['REMOTE_ADDR']  # first line added
        manager = getattr(instance, self.manager_name)

        pre_create_historical_record.send(
            sender=manager.model, instance=instance,
            history_date=history_date, history_user=history_user,
            history_change_reason=history_change_reason,
        )

        attrs = {}
        for field in self.fields_included(instance):
            attrs[field.attname] = getattr(instance, field.attname)
        history_instance = manager.create(
            ip_address=ip_address,  # second line added
            history_date=history_date, history_type=history_type,
            history_user=history_user,
            history_change_reason=history_change_reason, **attrs
        )

        post_create_historical_record.send(
            sender=manager.model, instance=instance,
            history_instance=history_instance,
            history_date=history_date, history_user=history_user,
            history_change_reason=history_change_reason,
)
```
As you can see I had to override the whole method.

2. Use the `post_create_historical_record` signal which results in an additional db hit.

**Additional context**
Just if somebody is confused. I have this base model:

```python
class IPAddressHistoricalModel(models.Model):
    """
    Abstract model for history models tracking the IP address.
    """
    ip_address = models.GenericIPAddressField(_('IP address'))

    class Meta:
        abstract = True
```

And I use it in my models like:
```python
    history = HistoricalRecords(bases=(IPAddressHistoricalModel,))
```

**Comment by rossmechanic** [comment]
The signals were released in 2.4.0 but I don't know why those would be backwards incompatible. So you want the history_instance so that you can add to it before the history_instance is saved? Essentially you just want to be able to add columns to the history instance (which you're doing with `bases`), but also add a callable that specifies how to populate those columns. Please correct me if I'm misunderstanding.

Going to add @kseever to this thread because he's currently working on adding a more generalizable way to add column/callable combinations to add to the history instance as he discussed in #422. I know we have a ticket for it that he's working on this week, so he should join in on this conversation. I think his next PR will solve your problem, but interested to see if this affects how he wants to do it.

**Comment by rossmechanic** [comment]
This was solves with #471. Closing this issue
